core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -0,0 +1,223 @@
+     */
+    IDLE,
+    /*
+     * We have scheduled a ping to be sent in the future. We may decide to delay it if we receive
+     * some data.
+     */
+    PING_SCHEDULED,
+    /*
+     * We need to delay the scheduled keepalive ping.
+     */
+    PING_DELAYED,
+    /*
+     * The ping has been sent out. Waiting for a ping response.
+     */
+    PING_SENT,
+    /*
+     * Transport goes idle after ping has been sent.
+     */
+    IDLE_AND_PING_SENT,
+    /*
+     * The transport has been disconnected. We won't do keepalives any more.
+     */
+    DISCONNECTED,
+  }
+
+  public KeepAliveManager(ManagedClientTransport transport, ScheduledExecutorService scheduler) {
+    this(transport, scheduler, SYSTEM_TICKER, KEEPALIVE_DELAY, KEEPALIVE_TIMEOUT);
+  }
+
+  @VisibleForTesting
+  KeepAliveManager(ManagedClientTransport transport, ScheduledExecutorService scheduler,
+                          Ticker ticker, long keepAliveDelayInNanos, long keepAliveTimeoutInNanos) {
+    this.transport = Preconditions.checkNotNull(transport, "transport");
+    this.scheduler = Preconditions.checkNotNull(scheduler, "scheduler");
+    this.ticker = Preconditions.checkNotNull(ticker, "ticker");
+    this.keepAliveDelayInNanos = keepAliveDelayInNanos;
+    this.keepAliveTimeoutInNanos = keepAliveTimeoutInNanos;
+    nextKeepaliveTime = ticker.read() + keepAliveDelayInNanos;
+  }
+
+  /**
+   * Transport has received some data so that we can delay sending keepalives.
+   */
+  public synchronized void onDataReceived() {
+    nextKeepaliveTime = ticker.read() + keepAliveDelayInNanos;
+    // We do not cancel the ping future here. This avoids constantly scheduling and cancellation in
+    // a busy transport. Instead, we update the status here and reschedule later. So we actually
+    // keep one sendPing task always in flight when there're active rpcs.
+    if (state == State.PING_SCHEDULED) {
+      state = State.PING_DELAYED;
+    }
+  }
+
+  /**
+   * Transport is actively sending data. Start sending keepalives if necessary.
+   */
+  public synchronized void onDataSent() {
+    if (state == State.IDLE) {
+      // When the transport goes active, we do not reset the nextKeepaliveTime. This allows us to
+      // quickly check whether the conneciton is still working.
+      state = State.PING_SCHEDULED;
+      pingFuture = scheduler.schedule(sendPing, nextKeepaliveTime - ticker.read(),
+          TimeUnit.NANOSECONDS);
+    }
+  }
+
+  /**
+   * Transport has finished all rpcs.
+   */
+  public synchronized void onTransportIdle() {
+    if (state == State.PING_SCHEDULED || state == State.PING_DELAYED) {
+      state = State.IDLE;
+    }
+    if (state == State.PING_SENT) {
+      state = State.IDLE_AND_PING_SENT;
+    }
+  }
+
+  class KeepAlivePingCallback implements ClientTransport.PingCallback {
+
+    @Override
+    public void onSuccess(long roundTripTimeNanos) {
+      synchronized (KeepAliveManager.class) {
```

Comment:
KeepAliveManager.this.

.class is synchronizing on the `Class` instance, which is really only done with static state.